proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,489 @@
+// Kotlin + KEEP-87
+fun <A> fetchById(id: String, with repository: Repository<A>): A? {
+  return loadById(id)
+}
+
+// Regular Kotlin
+fun <A> fetchById(id: String, repository: Repository<A>): A? =
+  with (semigroup) {
+    return loadById(id)
+  }
+```
+
+On the call site, we could use it like:
+
+```kotlin
+fetchById<User>("1182938") // compiles since we got evidence of a `Repository<User>` in scope.
+fetchById<Coin>("1239821") // does not compile: No `Repository<Coin>` evidence defined in scope!
+```
+
+Functions with extension parameters can be passed all values, or extension ones can be omitted and let the compiler resolve the suitable extensions for them. That makes the approach really similar to how default arguments work in Kotlin.
+
+```kotlin
+fetchById<User>("1182938") // compiles since we got evidence of a `Repository<User>` in scope.
+fetchById<User>("1182938", UserRepository()) // you can provide it manually.
+```
+
+When `with` is used in class constructors, it is important to **add val to extension class fields** to make sure they are accessible in the scope of the class. Here, the `with` keyword adds the value to the scope of every method in the class. In this scenario, the following classes would be equivalent:
+
+```kotlin
+data class Group<A>(val values: List<A>)
+
+// Kotlin + KEEP-87
+extension class GroupRepository<A>(with val R: Repository<A>) : Repository<Group<A>> {
+  override fun loadAll(): Group<A> =
+    Group(R.loadAll())
+}
+
+// Regular Kotlin
+class GroupRepository<A>(val R: Repository<A>) : Repository<Group<A>> {
+  override fun loadAll(): Group<A> = with (R) {
+    Group(loadAll())
+  }
+}
+```
+
+## Composition and chain of evidences
+
+Constraint interface declarations and extension evidences can encode further constraints on their type parameters so that they can be composed nicely:
+
+```kotlin
+package com.data.instances
+
+import com.data.Repository
+import com.domain.User
+import com.domain.Group
+
+extension class GroupRepository<A>(with val repoA: Repository<A>) : Repository<Group<A>> {
+  override fun loadAll(): List<Group<A>> {
+    return listOf(Group(repoA.loadAll()))
+  }
+
+  override fun loadById(id: Int): Group<A>? {
+    return Group(repoA.loadById(id))
+  }
+}
+```
+
+The above extension provides evidence of a `Repository<Group<A>>` as long as there is a `Repository<A>` in scope. Call site would be like:
+
+```kotlin
+fun fetchGroup<A>(with repo: GroupRepository<A>) = repo.loadAll()
+
+fun main() {
+  fetchGroup<User>() // Succeeds! There's evidence of Repository<Group<A>> and Repository<User> provided in scope.
+  fetchGroup<Coin>() // Fails! There's evidence of Repository<Group<A>> but no evidence of `Repository<Coin>` available.
+}
+```
+
+We believe the encoding proposed above fits nicely with Kotlin's philosophy of extensions and will reduce the boilerplate compared to other langs that also support compile time dependency resolution such as Scala where this is done via implicits.
```

Comment:
I'd drop the comparison to implicits here. It requires knowing how they work, and it doesn't give much. Maybe even the whole paragraph now that the proposal is already a body of work.

Reply:
Removed in 6ba2577.